lib/render/leaderboard.php

```diff
@@ -471,19 +475,50 @@ function getGlobalRankingData($lbType, $sort, $date, $user, $friendsOf = null, $
         default => "",
     };
 
-    // Run the All-Time ranking query
+    if ($unlockMode == UnlockMode::Hardcore) {
+        $totalAwards = "SUM(IF(AwardDataExtra > 0, 1, 0))";
+    } else {
+        $totalAwards = "COUNT(*)";
+        $pointRequirement = "AND ua.RASoftcorePoints >= 0"; // if someone resets a softcore achievement without resetting the hardcore, the query can return negative points
+    }
+
     $retVal = [];
-    if ($lbType == 2) {
+    if ($lbType == 2) { // Run the All-Time ranking query
+        if ($friendsOf === null) {
+            // if not comparing against friends, only look at the ranked users
+            if ($unlockMode == UnlockMode::Softcore) {
+                $pointRequirement = "AND ua.RASoftcorePoints >= " . Rank::MIN_POINTS;
+            } elseif ($sort == 6) {
+                $pointRequirement = "AND ua.TrueRAPoints >= " . Rank::MIN_TRUE_POINTS;
+            } else {
+                $pointRequirement = "AND ua.RAPoints >= " . Rank::MIN_POINTS;
+            }
+        }
+
         if ($info == 0) {
-            $selectQuery = "SELECT ua.User,
-                    (SELECT COALESCE(SUM(CASE WHEN HardcoreMode = " . UnlockMode::Hardcore . " THEN 1 ELSE 0 END), 0) FROM Awarded AS aw WHERE aw.User = ua.User) AS HardcoreCount,
-                    COALESCE(ua.RAPoints, 0) AS HardcorePoints,
-                    COALESCE(ua.TrueRAPoints, 0) AS RetroPoints,
-                    COALESCE(ROUND(ua.TrueRAPoints/ua.RAPoints, 2), 0) AS RetroRatio ";
+            if ($unlockMode == UnlockMode::Hardcore) {
+                $selectQuery = "SELECT ua.User,
+                        (SELECT COALESCE(SUM(CASE WHEN HardcoreMode = " . UnlockMode::Hardcore . " THEN 1 ELSE 0 END), 0) FROM Awarded AS aw WHERE aw.User = ua.User) AS AchievementCount,
+                        COALESCE(ua.RAPoints, 0) AS Points,
+                        COALESCE(ua.TrueRAPoints, 0) AS RetroPoints,
+                        COALESCE(ROUND(ua.TrueRAPoints/ua.RAPoints, 2), 0) AS RetroRatio ";
+            } else {
+                $selectQuery = "SELECT ua.User,
+                        (SELECT COALESCE(SUM(CASE WHEN HardcoreMode = " . UnlockMode::Softcore . " THEN 1 ELSE -1 END), 0) FROM Awarded AS aw WHERE aw.User = ua.User) AS AchievementCount,
```

Comment:
I think we may have to live with this behavior until we have some of the v2 data structures in place.

We can accurately count hardcore achievements for a date because we only have to look at the hardcore unlock and don't care about the softcore unlock. To accurately count softcore unlocks we have to join back to the Awarded table and only count entries where there isn't a hardcore unlock for the same achievement on _any day_. The v2 data structure will have both the hardcore unlock date and softcore unlock date in the same row so we can do the filters more efficiently.

For now, I'm going to add a filter to ignore Achievement counts less than zero (there's already a filter for Points less than zero for a similar reason).

Reply:
Yeah I don't think there is any reasonable way to do this accurately at the moment like you said, we can just make an issue for it to fix once the v2 structure is available.